arch: make more arch files kconfigurable

hakonfam opened the pull request:
Make more arch specific files kconfigurable so that the user can opt
out from using them. The motivation for this change is to get better
support for bare metal solutions.

Names shared between the arches is put in the top level kconfig.

Signed-off-by: Håkon Øye Amundsen <haakon.amundsen@nordicsemi.no>

galak reviewed (changes requested):
I think this needs more thought, and isn't the proper granularity.  It creates a slew of config options / matrix that we might have to test/validate.

hakonfam committed:
arch: make more arch files kconfigurable

Make more arch specific files kconfigurable so that the user can opt
out from using them. The motivation for this change is to get better
support for bare metal solutions.

Names shared between the arches is put in the top level kconfig.

Signed-off-by: Håkon Øye Amundsen <haakon.amundsen@nordicsemi.no>

andrewboie commented on arch/Kconfig:
```diff
@@ -71,6 +71,49 @@ module = MPU
+	help
+	  Include arch specific handlers for fatal errors/faults.
+
+config PREP_C
```
Why do we need this????
This is mandatory code to even boot....

tyrantlizard77 reviewed (changes requested):
Why do we need this?
What problem is this solving?
This needs an RFC, some kind of justification, right now looking at this patch I'm just confused. I'm fully willing to give the benefit of the doubt, but I'd like to understand well the problem space.

aescolar commented:
@hakonfam Would your problem be fixed with an out of tree architecture?

aescolar commented:
> I don't understand what you are proposing

@hakonfam The build system allows you to set ARCH_ROOT (and SOC_ROOT, and BOARD_ROOT) to define a path to an out of tree folder where you would define a new architecture (or SOC or BOARD). If you are developing a variant or new ARCH you want to keep for yourself, you should be able to do so by keeping it in its own repository and pointing to it with ARCH_ROOT.

hakonfam commented:
Thanks for clarifying!

I will see if that solution works better, even with a mix of custom files and ustream zephyr files.

hakonfam commented:
Using out of tree architecture does not solve the problem, as several of the arch-files will be used in the bare metal solution, ie `arch/${ARCH}/include` and `arch/${ARCH}/core/offsets/`. As there are pointed to directly by the root `CMakeLists.txt` it is not possible to tell the build system to use the upstream variants for these files.

aescolar commented:
> as several of the arch-files will be used 

@hakonfam Care to submit a fix?

hakonfam commented:
Do you mean a fix which moves the inclusion of those files out of `zephyr/CMakeLists.txt` and in to all the `zephyr/arch/${ARCH}/CMakeLists.txt` files? IMHO it is good to have it in the common place as it is required for all arches to define those files.

The main reason for us not going for the out-of-tree arch is that we don't have an out-of-tree arch, we have a normal arm cortex-m arch. The desired functionality is not re-writing arch-code to make it fit a different arch, it is re-writing arch code to make it work better for "bare metal" (i.e. first stage bootloader) applications.

hakonfam commented:
@andrewboie I believe some more info has been supplied in the comments above. If you could give me your thoughts that would be much appreciated.

Note that the difference in code size we see is around 750 bytes. 

I can provide data on exactly where those come from if we choose to go further with this solution.

andrewboie commented:
> Note that the difference in code size we see is around 750 bytes.

Most of this stuff ought to be simply removed by gc-sections if unused.
I'd like to understand better where this is failing here.
